File: apps/api/app/schemas/admin_analytics.py
"""
Admin Analytics Schemas
"""
from pydantic import BaseModel
from typing import List, Dict, Optional, Any
from datetime import date, datetime


# ============ Overview Endpoint ============

class AdminAnalyticsOverview(BaseModel):
    """Real-time snapshot of all key metrics"""
    # User counts
    total_users: int
    active_users_today: int
    active_users_7d: int
    active_users_30d: int

    # User growth
    new_users_today: int
    new_users_7d: int
    new_users_30d: int
    user_growth_rate: float  # Percentage change vs previous period

    # Revenue/Premium
    total_premium_users: int
    mrr: float  # Monthly Recurring Revenue
    subscription_conversion_rate: float  # Percentage of users on premium

    # Platform activity
    total_tarantulas: int
    total_feedings_today: int
    total_molts_today: int
    total_substrate_changes_today: int

    # Community
    total_forum_threads: int
    total_forum_posts: int
    total_messages_today: int

    # Timestamp
    generated_at: datetime


# ============ User Analytics ============

class UserTimeSeriesPoint(BaseModel):
    """Single data point for user time series"""
    date: date
    new_users: int
    cumulative_users: int
    active_users: int


class UserRetentionCohort(BaseModel):
    """Retention data for a cohort of users"""
    cohort_month: str  # e.g., "2024-01"
    cohort_size: int
    retained_day_1: float
    retained_day_7: float
    retained_day_30: float


class OAuthBreakdown(BaseModel):
    """OAuth provider breakdown"""
    provider: str
    count: int
    percentage: float


class UserAnalyticsResponse(BaseModel):
    """Response for user analytics endpoint"""
    time_series: List[UserTimeSeriesPoint]
    retention_cohorts: List[UserRetentionCohort]
    oauth_breakdown: List[OAuthBreakdown]

    # Summary stats for period
    total_new_users: int
    growth_rate: float
    average_retention_day_30: float


# ============ Revenue Analytics ============

class RevenueTimeSeriesPoint(BaseModel):
    """Single data point for revenue time series"""
    date: date
    new_subscriptions: int
    cancellations: int
    active_subscriptions: int


class SubscriptionBreakdown(BaseModel):
    """Subscription plan breakdown"""
    plan_name: str
    active_count: int
    percentage: float


class SubscriptionSourceBreakdown(BaseModel):
    """Subscription source breakdown"""
    source: str  # 'promo', 'monthly', 'yearly', 'lifetime', 'iap'
    count: int
    percentage: float


class RevenueAnalyticsResponse(BaseModel):
    """Response for revenue analytics endpoint"""
    time_series: List[RevenueTimeSeriesPoint]
    subscription_breakdown: List[SubscriptionBreakdown]
    source_breakdown: List[SubscriptionSourceBreakdown]

    # Key metrics
    mrr: float
    total_premium_users: int
    churn_rate: float  # Cancellations / total over period
    conversion_rate: float  # Premium users / total users


# ============ Activity Analytics ============

class ActivityTimeSeriesPoint(BaseModel):
    """Single data point for activity time series"""
    date: date
    new_tarantulas: int
    feedings: int
    molts: int
    substrate_changes: int


class SpeciesPopularity(BaseModel):
    """Species popularity data"""
    species_name: str
    count: int
    percentage: float


class ActivityAnalyticsResponse(BaseModel):
    """Response for activity analytics endpoint"""
    time_series: List[ActivityTimeSeriesPoint]
    top_species: List[SpeciesPopularity]

    # Summary stats
    total_tarantulas: int
    total_feedings: int
    total_molts: int
    average_collection_size: float


# ============ Community Analytics ============

class CommunityTimeSeriesPoint(BaseModel):
    """Single data point for community time series"""
    date: date
    new_threads: int
    new_posts: int
    new_messages: int
    new_follows: int


class ForumCategoryStats(BaseModel):
    """Forum category statistics"""
    category_name: str
    thread_count: int
    post_count: int


class TopContributor(BaseModel):
    """Top contributor data"""
    user_id: str
    username: str
    display_name: Optional[str]
    post_count: int


class CommunityAnalyticsResponse(BaseModel):
    """Response for community analytics endpoint"""
    time_series: List[CommunityTimeSeriesPoint]
    forum_categories: List[ForumCategoryStats]
    top_contributors: List[TopContributor]

    # Summary stats
    total_threads: int
    total_posts: int
    total_follows: int
    average_posts_per_thread: float

File: apps/web/src/types/admin-analytics.ts
/**
 * Admin Analytics TypeScript Types
 * Matches the backend Pydantic schemas
 */

export type AnalyticsPeriod = '7d' | '30d' | '90d' | '1y';

// ============ Overview ============

export interface AdminAnalyticsOverview {
  total_users: number;
  active_users_today: number;
  active_users_7d: number;
  active_users_30d: number;
  new_users_today: number;
  new_users_7d: number;
  new_users_30d: number;
  user_growth_rate: number;
  total_premium_users: number;
  mrr: number;
  subscription_conversion_rate: number;
  total_tarantulas: number;
  total_feedings_today: number;
  total_molts_today: number;
  total_substrate_changes_today: number;
  total_forum_threads: number;
  total_forum_posts: number;
  total_messages_today: number;
  generated_at: string;
}

// ============ User Analytics ============

export interface UserTimeSeriesPoint {
  date: string;
  new_users: number;
  cumulative_users: number;
  active_users: number;
}

export interface UserRetentionCohort {
  cohort_month: string;
  cohort_size: number;
  retained_day_1: number;
  retained_day_7: number;
  retained_day_30: number;
}

export interface OAuthBreakdown {
  provider: string;
  count: number;
  percentage: number;
}

export interface UserAnalyticsResponse {
  time_series: UserTimeSeriesPoint[];
  retention_cohorts: UserRetentionCohort[];
  oauth_breakdown: OAuthBreakdown[];
  total_new_users: number;
  growth_rate: number;
  average_retention_day_30: number;
}

// ============ Revenue Analytics ============

export interface RevenueTimeSeriesPoint {
  date: string;
  new_subscriptions: number;
  cancellations: number;
  active_subscriptions: number;
}

export interface SubscriptionBreakdown {
  plan_name: string;
  active_count: number;
  percentage: number;
}

export interface SubscriptionSourceBreakdown {
  source: string;
  count: number;
  percentage: number;
}

export interface RevenueAnalyticsResponse {
  time_series: RevenueTimeSeriesPoint[];
  subscription_breakdown: SubscriptionBreakdown[];
  source_breakdown: SubscriptionSourceBreakdown[];
  mrr: number;
  total_premium_users: number;
  churn_rate: number;
  conversion_rate: number;
}

// ============ Activity Analytics ============

export interface ActivityTimeSeriesPoint {
  date: string;
  new_tarantulas: number;
  feedings: number;
  molts: number;
  substrate_changes: number;
}

export interface SpeciesPopularity {
  species_name: string;
  count: number;
  percentage: number;
}

export interface ActivityAnalyticsResponse {
  time_series: ActivityTimeSeriesPoint[];
  top_species: SpeciesPopularity[];
  total_tarantulas: number;
  total_feedings: number;
  total_molts: number;
  average_collection_size: number;
}

// ============ Community Analytics ============

export interface CommunityTimeSeriesPoint {
  date: string;
  new_threads: number;
  new_posts: number;
  new_messages: number;
  new_follows: number;
}

export interface ForumCategoryStats {
  category_name: string;
  thread_count: number;
  post_count: number;
}

export interface TopContributor {
  user_id: string;
  username: string;
  display_name: string | null;
  post_count: number;
}

export interface CommunityAnalyticsResponse {
  time_series: CommunityTimeSeriesPoint[];
  forum_categories: ForumCategoryStats[];
  top_contributors: TopContributor[];
  total_threads: number;
  total_posts: number;
  total_follows: number;
  average_posts_per_thread: number;
}
